aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+   * this will call an unboxing wrapper which then calls into that
+   * unboxed function.
+   *
+   * Example:
+   *
+   * > void boxed_func(OperatorKernel*, Stack* stack) {...}
+   * > KernelFunction func = KernelFunction::makeFromBoxedFunction(&boxed_func);
+   * > Tensor result = func.callBoxed(stack);
+   *
+   * Or, with an unboxed implementation:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callBoxed(stack);
+   */
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitizliaed KernelFunction.");
+      } else {
+        // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this case should be impossible.
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on a KernelFunction that can only be called with KernelFunction::callUnboxed().");
+      }
+    }
+
+    (*boxed_kernel_func_)(getFunctor_(), stack);
+  }
+
+  /**
+   * Call the function in an unboxed way.
+   * As the "Only" in the name suggests, this only works for KernelFunctions
+   * that are backed by an unboxed kernel. If the KernelFunction was created
+   * in a boxed way, this will fail (also see KernelFunction::callUnboxed()).
+   *
+   * KernelFunction::callUnboxed() is generally better, since it will allow
+   * calling KernelFunctions that are backed by either boxed or unboxed
+   * kernels, but that one will not work for all types.
+   *
+   * Example:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callUnboxedOnly<Tensor, Tensor, bool>(tensor1, true);
+   */
+  template<class Return, class... Args>
+  Return callUnboxedOnly(Args... args) const {
+    // note: Args above is intentionally not Args&&. We don't want perfect
+    // forwarding, which would require Args to be deduced, but instead we
+    // want callers to explicitly specify the Args.
+
+    // TODO Remove this function once all kernels support a boxed variant
+
+    TORCH_INTERNAL_ASSERT(!signature_hash_.has_value() || (detail::hashFunctionSignature<Return (Args...)>() == *signature_hash_),
+      "Called KernelFunction::callUnboxedOnly with wrong argument types");
+
+    if (unboxed_kernel_func_ != nullptr) {
+      using ActualSignature = Return (OperatorKernel*, Args...);
+      ActualSignature* func = reinterpret_cast<ActualSignature*>(unboxed_kernel_func_);
+      return (*func)(getFunctor_(), std::forward<Args>(args)...);
+    }
+
+    TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callUnboxedOnly() for a kernel that doesn't have an unboxed version.");
+  }
+
+  /**
+   * Call the function in an unboxed way.
```

Comment:
At risk of sounding like a broken record, this API is sooooooo big. ATenDispatch had only a single way of registering a function and it made working with the code a lot easier. Doe we *really* need all these ways of registering functions? Maybe in an ideal world we would have all of them, but can we do without some, without too much trouble?

Reply:
Technically, the API surface is smaller than it looks like. Most of these factory functions just forward to the functor one, they don't access any class internals. So technically, the API surface for creating KernelFunction is only two methods, one for creating from a boxed function and one for creating from an unboxed functor.

Reply:
Note also that we need functors + functions + lambdas because they're part of the public custom op API. Some of them (the `unboxedOnly` variants) are needed only temporarily while we're doing the ATen->c10 transition and can be deleted later.

Reply:
> Technically, the API surface is smaller than it looks like. Most of these factory functions just forward to the functor one, they don't access any class internals. So technically, the API surface for creating KernelFunction is only two methods, one for creating from a boxed function and one for creating from an unboxed functor.

The big step in complexity is having the internal implementation implement functors instead of just function pointers. We've given a lot of evidence that boxed functors are not actually necessary, *except* for the case when you are autogenerating a boxed wrapper for a unboxed function, but even then, rather than a general functor concept, a simple, NON-user-visible API for this use case would have been sufficient.

> Note also that we need functors + functions + lambdas because they're part of the public custom op API

There isn't that much usage yet. That is even more of an argument we should remove functionality now, before it becomes a big deal to remove it later. For reference, https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html#registering-the-custom-operator-with-torchscript only advertises a *single* way of registering operators. Can you imagine having to explain three different APIs to users, and when to use them? It's totally unnecessary.

Reply:
This PR just refactors things into a better state, it doesn't change any functionality or semantics and I won't change functionality in here, the PR is already complex enough to understand when you know it just refactors. If you want to change the public API, that's a different discussion we need to have, but that is unrelated to this PR.